### ISSUE TITLE
RF train_flag_ctx potentially not what you want, enable_dropout_ctx or enable_regularization_ctx or so instead

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1712]
When doing forwarding/recog, and I want to enable dropout there (e.g. to get some variance into my outputs), I used to do this:
```python
with rf.get_run_ctx().train_flag_ctx(True):
    ...
```
This worked. Now dropout is used inside this block even though this is in forward/recog.

However, I realized that this can have some unwanted other side effects, depending on the model, whenever sth else is also dependent on the train flag. In my case, this was actually the case, as I had `BatchNorm` in there. So with this code, instead of using the learned batch norm stats, it was using the on-the-fly calculated batch norm stats (and also updating the saved batch norm stats, although that was never used then).

So, I think the best solution is really to avoid using `train_flag_ctx` for such intents, and instead introduce a new flag, specifically for dropout, or maybe also for other regularization in general. In my current use case, I actually also want to enable specaugment. So there could maybe also be a enable_regularization flag or so. Functions like dropout and specaugment would first check for that flag, and if this is unset, fallback to the train_flag.

(cc @dorian-K)

**Comment by albertz** [comment]
@NeoLegends any suggestions?

**Comment by NeoLegends** [comment]
We could also structure the models to accept individual flags for each runtime-dependent feature, that then default to whether you are in the train flag or not so that the entire feature is not too verbose.

**Comment by albertz** [comment]
I'm not sure how exactly you mean that. Similar to the `training` flag in PyTorch? But what exactly do you mean then? That we have `training` in every RF module? But how is that a solution to the problem here? It would just change the code to this (following exactly the PyTorch API):
```python
model.train(True)
```
and have exactly the same problem then? I want to do sth like this but only for selected methods, like dropout / specaugment.

Also note, I cannot just iterate through `model.modules()` and check for `Dropout`/`SpecAugment` modules, because dropout/specaugment is implemented and used in a functional way, i.e. there is no module anywhere for it. (And even if there would be such modules, it could be that this is not used consistently everywhere. E.g. in PyTorch, you could also use `torch.nn.functional.dropout1d` or so, and then such logic would also not work.)

I think there is no way around a global flag, which could be controlled via a context manager, just like `train_flag_ctx` is doing it already.

So, the only question is, how should this global flag look like? Should there be a flag specifically only for dropout, and then a separate flag e.g. for specaugment, etc? Or should there be a flag for all kinds of regularizations, including dropout, specaugment, layer drop, and similar things? Or other suggestions?

**Comment by albertz** [comment]
Another suggestion for more fine-grained control:

Maybe you still don't want to enable all kind of regularization, but really only some selected specific one, like dropout and specaugment, but nothing else. (This is actually my use case currently.) The suggestion here for a regularization flag would already be a big improvement over the train flag, but there could still be cases where this would behave wrong.

There could be some API like:

```python
@contextmanager
def regularization_ctx(enabled: bool, *, func: Optional[FunctionType] = None): ...
```

You could use this like:
```python
with rf.regularization_ctx(True, func=rf.dropout):
    ...
```

That would enable regularization only for `rf.dropout`. (For other functions, it would derive it from the parent scope. If not defined, it would fallback to train_flag.)

Or the global case:
```python
with rf.regularization_ctx(True):
    ...
```
That would enable it for any function.

**Comment by albertz** [comment]
I'm actually thinking, when it is generalized that way, we could also keep the old name, i.e. extend `RunCtx.train_flag_ctx`, i.e. like this:
```python
class RunCtx:

    ...

    @contextmanager
    def train_flag_ctx(self, train_flag: Union[bool, Tensor], *, func: Optional[FunctionType] = None):
        ...

    def is_train_flag_enabled(self, *, func: Optional[FunctionType] = None):
        ...
```
